cranelift/codegen/src/machinst/abi.rs

```diff
@@ -2029,19 +2087,51 @@ impl<M: ABIMachineSpec> Caller<M> {
         }
     }
 
-    /// Generate a copy of an argument value from a source register, prior to
-    /// the call.  For large arguments with associated stack buffer, this may
-    /// load the address of the buffer into the argument register, if required
-    /// by the ABI.
-    pub fn gen_copy_regs_to_arg(
-        &self,
-        ctx: &Lower<M::I>,
+    /// Add a constraint for an argument value from a source register.
+    /// For large arguments with associated stack buffer, this may
+    /// load the address of the buffer into the argument register, if
+    /// required by the ABI.
+    pub fn gen_arg(
+        &mut self,
+        ctx: &mut Lower<M::I>,
         idx: usize,
         from_regs: ValueRegs<Reg>,
     ) -> SmallInstVec<M::I> {
         let mut insts = smallvec![];
         let word_rc = M::word_reg_class();
         let word_bits = M::word_bits() as usize;
+
+        // How many temps do we need for extends? Allocate them ahead
+        // of time, since we can't do it while we're iterating over
+        // the sig and immutably borrowing `ctx`.
```

Comment:
We could allow taking the signatures out of the context temporarily, so that we could immutably borrow them while mutably borrowing the lowering context. This would avoid needing to pre-allocate temps up front.

Reply:
True, though imho that is probably best done as a follow-up refactor...